### PR TITLE
[WB-9965] Add Job, QueuedRun and job handling in launch

### PR DESCRIPTION
**Pull request by KyleGoyette**
[Fixes WB-9965](https://wandb.atlassian.net/browse/WB-9965)
Fixes #NNNN

Description
-----------
- adds the Job api
- adds the QueuedRun api
- modifies launch, and launch_add to use these properly
See master PR for description: https://github.com/wandb/client/pull/3675

Will require: https://github.com/wandb/yea-wandb/pull/69 to merge before yea-sync will pass

Testing
-------
How was this PR tested?

Checklist
-------
- Include reference to internal ticket "Fixes WB-NNNN" (and github issue "Fixes #NNNN" if applicable)

### CODE REVIEW

File: tests/tests_launch/test_launch.py [hunk 1]
```diff
@@ -179,6 +179,15 @@ def mock_cuda_run_info(monkeypatch):
     )
 
 
+def code_download_func(dst_dir):
+    with open(os.path.join(dst_dir, "train.py"), "w") as f:
```
Comment (hu-po) [inline tests/tests_launch/test_launch.py:183]: maybe wrap this block with a `try` in case these files aren't found and then throw a LaunchError?

File: tests/tests_launch/test_launch.py [hunk 2]
```diff
@@ -332,9 +344,20 @@ def test_launch_resource_args(
...
-def test_launch_add_base(live_mock_server):
-    queuedJob = launch_add("https://wandb.ai/mock_server_entity/tests/runs/1")
-    assert queuedJob._run_queue_item_id == "1"
+def test_launch_add_base_queued_run(live_mock_server):
+    queued_run = launch_add("https://wandb.ai/mock_server_entity/tests/runs/1")
+    assert queued_run.state == "pending"
+    assert queued_run.run_queue_item_id == "1"
+    assert queued_run.entity == "mock_server_entity"
+    assert queued_run.project == "tests"
+    assert queued_run.username == "mock_server_entity"
```
Comment (hu-po) [inline tests/tests_launch/test_launch.py:353]: Lol, I still don't really understand why `username` versus `entity` are different

File: tests/tests_launch/test_launch_jobs.py [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+            lambda *arg, **kwargs: artifact_fetcher(*arg, **kwargs),
+        )
+
+    return mock_artifact_fetcher
+
+
+def test_fetch_job_fail(api):
+
+    launch_project = _project_spec.LaunchProject(
+        None,
```
Comment (hu-po) [inline tests/tests_launch/test_launch_jobs.py:64]: This is where the explicit kwarg pattern would be nice. I have no idea what all these different things are.

File: wandb/cli/cli.py [hunk 4]
```diff
@@ -1093,17 +1101,20 @@ def launch(
     if (
         uri is None
         and docker_image is None
-        and config.get("uri") is not None
+        and job is None
+        and config.get("uri") is None
+        and config.get("job") is None
```
Comment (hu-po) [inline wandb/cli/cli.py:1106]: is git also an option?

File: wandb/sdk/launch/launch.py [hunk 5]
```diff
@@ -167,8 +169,9 @@ def _run(
 
 
 def run(
-    uri: Optional[str],
     api: Api,
+    uri: Optional[str] = None,
+    job: Optional[str] = None,
```
Comment (hu-po) [inline wandb/sdk/launch/launch.py:174]: Typed kwargs ☺️

File: wandb/sdk/launch/launch_add.py [hunk 6]
```diff
@@ -17,7 +17,8 @@ def push_to_queue(api: Api, queue: str, launch_spec: Dict[str, Any]) -> Any:
 
 
 def launch_add(
-    uri: str,
+    uri: Optional[str] = None,
```
Comment (hu-po) [inline wandb/sdk/launch/launch_add.py:20]: You could just simplify this passthrough func and make it just pass forward `*args, **kwargs`? If they are named kwargs then you don't even need to worry about order.

File: wandb/sdk/launch/launch_add.py [hunk 7]
```diff
@@ -91,14 +101,28 @@ def _launch_add(
         launch_config,
         cuda,
     )
-
+    if (
+        launch_spec.get("uri") is None
+        and launch_spec.get("job") is None
+        and launch_spec.get("docker", {}).get("docker_image") is None
+    ):
+        raise ValueError("Must specify either uri or job or docker_image")
```
Comment (hu-po) [inline wandb/sdk/launch/launch_add.py:109]: I feel like this gets done enough throughout the PR that you could add a utility called `validate_launch_spec` or something like that, which performs this check. You could then also create a unit test for it and save yourself all this code cov spam.

File: wandb/util.py [hunk 8]
```diff
@@ -1695,3 +1695,13 @@ def ensure_text(
         return string
     else:
         raise TypeError(f"not expecting type '{type(string)}'")
+
+
+def make_artifact_name_safe(name: str) -> str:
+    """Make an artifact name safe for use in artifacts"""
+    return re.sub(r"[^a-zA-Z0-9_\-.]", "", name)
```
Comment (hu-po) [inline wandb/util.py:1702]: lolwut

**Review by hu-po — COMMENTED** [review]
Mostly just code style/pattern comments. Prolly worth getting someone who understands the system and design (especially the artifact stuff) to look over this.

File: tests/tests_launch/test_launch.py (hunk 2 above)
Comment (KyleGoyette) [reply tests/tests_launch/test_launch.py:353]: My understanding is that we used to allow users to refer to username, but we've switched to the entity paradigm, I can remove this check

File: wandb/util.py (hunk 8 above)
Comment (KyleGoyette) [reply wandb/util.py:1702]: Adding a comment to make it clear

File: wandb/sdk/launch/launch_add.py (hunk 7 above)
Comment (KyleGoyette) [reply wandb/sdk/launch/launch_add.py:109]: yeah, I'll put it on LaunchProject init, and remove other references to them

Comment (KyleGoyette) [reply wandb/sdk/launch/launch_add.py:109]: Oo actually it will need to be on the spec. okay!

File: wandb/sdk/launch/launch_add.py (hunk 6 above)
Comment (KyleGoyette) [reply wandb/sdk/launch/launch_add.py:20]: We should add documentation to this and leave this as is, in my opinion.

File: wandb/cli/cli.py (hunk 4 above)
Comment (KyleGoyette) [reply wandb/cli/cli.py:1106]: yes it is, I'll get rid of this and put it in the spec validation function.

File: tests/utils/artifact_emu.py [hunk 9]
```diff
@@ -110,13 +110,15 @@ def create_files(self, variables):
     def query(self, variables, query=None):
         public_api_query_str = "query Artifact($id: ID!) {"
         public_api_query_str2 = "query ArtifactWithCurrentManifest($id: ID!) {"
+        public_api_query_str3 = "query ArtifactManifest("
         art_id = variables.get("id")
         art_name = variables.get("name")
         assert art_id or art_name
```
[the remark this answers is not in the document]
Comment (KyleGoyette) [reply tests/utils/artifact_emu.py:116]: I should have been more explicit here. This file is being reviewed in this pr: https://github.com/wandb/client/pull/3807

File: tests/tests_launch/test_launch.py (hunk 1 above)
Comment (hu-po) [reply tests/tests_launch/test_launch.py:183]: Yes you are correct, I think in my scrolling frenzy I missed that this is inside a test.

**Review by hu-po — APPROVED** [review]
The release of this PR was prophesied in the ancient texts.

File: wandb/apis/public.py [hunk 10]
```diff
@@ -2136,6 +2087,458 @@ def __repr__(self):
...
+        # TODO: allow users to update run queue items using queued run
+        if self._run is None:
+            raise ValueError(f"{self._run_required_error_message}")
+        else:
+            self._run.update()
+
+    @normalize_exceptions
+    def delete(self, delete_artifacts=False):
+        """
+        Deletes the given run from the wandb backend.
```
Comment (tssweeney) [inline wandb/apis/public.py:2244]: nit: "queued run"

File: wandb/apis/public.py [hunk 11]
```diff
@@ -2136,6 +2087,458 @@ def __repr__(self):
         return "<Run {} ({})>".format("/".join(self.path), self.state)
 
 
+class QueuedRun(Attrs):
```
Comment (tssweeney) [inline wandb/apis/public.py:2090]: Seems like you are doing a lot of work to proxy method calls to the underlying run object. this is going to be hard to maintain. Could we just have a getter that returns the underlying run object?

File: wandb/apis/public.py [hunk 12]
```diff
@@ -4441,3 +4844,128 @@ def convert_objects(self):
 
     def __repr__(self):
         return "<ArtifactFiles {} ({})>".format("/".join(self.artifact.path), len(self))
+
+
+class Job(Media):
```
Comment (tssweeney) [inline wandb/apis/public.py:4849]: Why is this a Media subclass? can you elaborate here?

File: wandb/apis/public.py [hunk 13]
```diff
@@ -4441,3 +4844,128 @@ def convert_objects(self):
...
+    _entrypoint: List[str]
+
+    def __init__(self, client: Api, name, path: str = None) -> None:
+
+        self._job_artifact = client.artifact(name, type="job")
+        if path:
+            self._fpath = path
+            self._job_artifact.download(root=path)
+        else:
+            self._fpath = self._job_artifact.download()
```
Comment (tssweeney) [inline wandb/apis/public.py:4866]: How are you guaranteed to have the corresponding artifact?

File: wandb/apis/public.py (hunk 12 above)
Comment (tssweeney) [reply wandb/apis/public.py:4849]: the ownership pattern here is a bit inverted w.r.t Media and Artifacts. an Artifact can have an entry that describes a Media object. However in this implementation, the constructor fetches an artifact. I think looking at how the Video media type works for example would be more akin to the desired pattern for media types

Comment (tssweeney) [reply wandb/apis/public.py:4849]: the `from_json` method does all the artifact member reading before passing to an artifact-agnostic constructor

File: wandb/sdk/launch/launch_add.py [hunk 14]
```diff
@@ -28,11 +29,56 @@ def launch_add(
...
     params: Optional[Dict[str, Any]] = None,
-) -> "public.QueuedJob":
+    resource_args: Optional[Dict[str, Any]] = None,
+    cuda: Optional[bool] = None,
+) -> "public.QueuedRun":
+    """Enqueue a W&B launch experiment.
+
+    Arguments:
+    uri: URI of experiment to run. A wandb run uri or a Git repository URI.
+    job: string reference to a wandb.Job eg: wandb/test/my-job:latest
```
Comment (tssweeney) [inline wandb/sdk/launch/launch_add.py:39]: is this supposed to be 1 or the other? can that be clarified?

**Review by tssweeney — CHANGES_REQUESTED** [review]
Just reviewing from a code perspective. The main concern here is the Media sublassing for Job. See comment.

File: wandb/apis/public.py (hunk 10 above)
Comment (KyleGoyette) [reply wandb/apis/public.py:2244]: Fixed

File: wandb/apis/public.py (hunk 13 above)
Comment (KyleGoyette) [reply wandb/apis/public.py:4866]: I'll wrap this in a try and raise an appropriate error

File: wandb/sdk/launch/launch_add.py (hunk 14 above)
Comment (KyleGoyette) [reply wandb/sdk/launch/launch_add.py:39]: Yep, added to the main description above. Also passing in more than one raises an error indicating this issue

File: tests/tests_launch/test_launch_jobs.py [hunk 15]
```diff
@@ -0,0 +1,231 @@
...
+        "Test_project",
+        None,
+        {},
+        {},
+        {},
+        "local",
+        {},
+        None,
+    )
+    with pytest.raises(LaunchError) as e_info:
```
[the remark this answers is not in the document]
Comment (KyleGoyette) [reply tests/tests_launch/test_launch_jobs.py:78]: lol I switch back and forth tbh

File: wandb/apis/public.py [hunk 16]
```diff
@@ -2136,6 +2087,183 @@ def __repr__(self):
...
+        }
+        res = self.client.execute(query, variable_values)
+        for item in res["project"]["runQueue"]["runQueueItems"]["edges"]:
+            if str(item["node"]["id"]) == str(self.id):
+                return item["node"]["state"].lower()
+        raise ValueError(
+            f"Could not find QueuedRun associated with id: {self.id} on queue id {self.queue_id}"
+        )
+
+    def _exec(self, query, **kwargs):
```
Comment (tssweeney) [inline wandb/apis/public.py:2163]: Why is this here? This seems like an odd place to put a general query layer

File: wandb/apis/public.py [hunk 17]
```diff
@@ -2136,6 +2087,183 @@ def __repr__(self):
...
+        variable_values = {
+            "projectName": self.project,
+            "entityName": self._entity,
+            "runQueue": self.queue_id,
+        }
+
+        while True:
+            res = self.client.execute(query, variable_values)
+            # TODO: add fetch run queue by item end point
+            for item in res["project"]["runQueue"]["runQueueItems"]["edges"]:
```
Comment (tssweeney) [inline wandb/apis/public.py:2240]: In the off-chance that the original query does not return a matching node, then i think we hit an inf loop? I think if the inner if block is not entered on the first iteration, we should error/break early

**Review by tssweeney — APPROVED** [review]
This also looks good! I left two comments, but not blocking.
